### PR TITLE
unix: make (get|set)_process_title MT-safe

### PR DESCRIPTION
**Pull request by mtharrison**
This is my attempt at fixing: https://github.com/libuv/libuv/issues/271

- I've gone for a mutex based on the conversation with @bnoordhuis in the issue
- This is my first contribution to libuv or any C project, open to all feedback especially on:
  - The best way to reduce all the repetition for the different platforms (e.g. init_once(void)` in each of the files)
  - How to write tests for this (I wrote this script to simulate issues and verify the fix: https://gist.github.com/mtharrison/e26e1db197d758479717b82f7849bb0f)
- I tried to make line spacing consistent but unsure what the rules are and if I made it worse or better

### CODE REVIEW

File: src/unix/aix.c [hunk 1]
```diff
@@ -64,11 +64,19 @@
...
+static uv_mutex_t mutex;
+static uv_once_t once = UV_ONCE_INIT;
 static void* args_mem = NULL;
 static char** process_argv = NULL;
 static int process_argc = 0;
 static char* process_title_ptr = NULL;
 
+
+static void init_once(void) {
+    uv_mutex_init(&mutex);
```
Comment (bnoordhuis) [inline src/unix/aix.c]: Style: two space indent.  (Also happens in other files.  I won't point them out individually.)

File: src/unix/aix.c [hunk 2]
```diff
@@ -946,12 +954,17 @@ char** uv_setup_args(int argc, char** argv) {
 int uv_set_process_title(const char* title) {
   char* new_title;
 
+  uv_once(&once, init_once);
+  uv_mutex_lock(&mutex);
```
Comment (bnoordhuis) [inline src/unix/aix.c]: You can move the `uv_mutex_lock()` call to right before the `if (process_title_ptr != NULL)` check.

File: src/unix/aix.c [hunk 3]
```diff
@@ -64,11 +64,19 @@
 #define RDWR_BUF_SIZE   4096
 #define EQ(a,b)         (strcmp(a,b) == 0)
 
+static uv_mutex_t mutex;
```
Comment (bnoordhuis) [inline src/unix/aix.c]: General observation: can you call this `process_title_mutex` or something like that?  Just `mutex` is a bit too generic.  Likewise for `once`.

File: src/unix/openbsd.c [hunk 4]
```diff
@@ -36,7 +36,15 @@
 #include <unistd.h>
 
 
+static uv_mutex_t mutex;
+static uv_once_t once = UV_ONCE_INIT;
 static char *process_title;
+static void* args_mem;
```
Comment (bnoordhuis) [inline src/unix/openbsd.c]: Unused variable.

File: src/unix/proctitle.c [hunk 5]
```diff
@@ -24,16 +24,26 @@
 #include <stdlib.h>
 #include <string.h>
 
+
```
Comment (bnoordhuis) [inline src/unix/proctitle.c]: Can you undo the whitespace changes in this file?

File: src/unix/proctitle.c [hunk 6]
```diff
@@ -84,10 +94,15 @@ int uv_set_process_title(const char* title) {
   if (process_title.len == 0)
     return 0;
 
+  uv_once(&once, init_once);
+  uv_mutex_lock(&mutex);
```
Comment (bnoordhuis) [inline src/unix/proctitle.c]: This needs to be moved to before the `process_title.len == 0` check.  I'd change it to this:
```c
uv_once(&once, init_once);
uv_mutex_lock(&mutex);
if (process_title.len != 0) {
  // ...
}
uv_mutex_unlock(&mutex);
return 0;
```

File: src/unix/proctitle.c [hunk 7]
```diff
@@ -98,11 +113,16 @@ int uv_get_process_title(char* buffer, size_t size) {
   else if (size <= process_title.len)
     return -ENOBUFS;
 
+  uv_once(&once, init_once);
+  uv_mutex_lock(&mutex);
```
Comment (bnoordhuis) [inline src/unix/proctitle.c]: Likewise, this needs to happen before the `size <= process_title.len` check.

**Comment by mtharrison** [comment]
Thanks for the feedback. I've made all the suggested changes. Just the test left to do, I'll add that tomorrow.

**Comment by mtharrison** [comment]
@cjihrig sure thing

**Comment by mtharrison** [comment]
Added a test in 583aa5ef794f3b66bdb35154c3c8078651b2afae. It took a bit of playing around with but this test consistently fails with master (OSX) for me and passes with my branch. With 1000 iterations the test was taking far too long to complete.

**Comment by mtharrison** [comment]
@cjihrig I noticed the warning about the functions not being thread safe is only present in the v1.x branch docs and not master which I'm making the PR against. Is there anything to be done in this case? Should I be making this PR against v1.x?

File: test/test-process-title-threadsafe.c [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+#include "uv.h"
+#include "task.h"
+
+#include <string.h>
+
+#define NUM_SETTER_THREADS  4
+#define NUM_ITERATIONS      100
+
+static char *titles[] = {
```
Comment (bnoordhuis) [inline test/test-process-title-threadsafe.c]: `const char` and star leans left.

File: test/test-process-title-threadsafe.c [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  "8L2NY0Kdj0XyNFZnmUZigIOfcWjyNr0SkMmUhKw99VLUsZFrvCQQC3XIRfNR8pjyMjXObllled",
+  "jUAcscJN49oLSN8GdmXj2Wo34XX2T2vp2j5khfajNQarlOulp57cE130yiY53ipJFnPyTn5i82",
+  "9niCI5icXGFS72XudhXqo5alftmZ1tpE7B3cwUmrq0CCDjC84FzBNB8XAHqvpNQfI2QAQG6ztT",
+  "n8qXVXuG6IEHDpabJgTEiwtpY6LHMZ8MgznnMpdHARu5EywufA6hcBaQfetb0YhEsK0ykDd7JU"
+};
+
+static void getter_thread_body(void* arg) {
+  char buffer[512];
+
+  while(1) {
```
Comment (bnoordhuis) [inline test/test-process-title-threadsafe.c]: Space before `(` but consider using `for (;;) {`, that's more idiomatic in libuv.

File: test/test-process-title-threadsafe.c [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+
+static void setter_thread_body(void* arg) {
+  int i;
+
+  for (i = 0; i < NUM_ITERATIONS; i++) {
+    ASSERT(uv_set_process_title(titles[0]) == 0);
+    ASSERT(uv_set_process_title(titles[1]) == 0);
+    ASSERT(uv_set_process_title(titles[2]) == 0);
+    ASSERT(uv_set_process_title(titles[3]) == 0);
```
Comment (bnoordhuis) [inline test/test-process-title-threadsafe.c]: Not wrong but we usually use Yoda style `0 == ...` in asserts.

File: test/test-process-title-threadsafe.c [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+  }
+}
+
+
+TEST_IMPL(process_title_threadsafe) {
+  uv_thread_t setter_threads[NUM_SETTER_THREADS];
+  uv_thread_t getter_thread;
+  int i;
+
+#if defined(__sun) || defined(__CYGWIN__) || defined(__MSYS__)
```
Comment (bnoordhuis) [inline test/test-process-title-threadsafe.c]: Also not implemented on `__MVS__` (os390.)

File: test/test-process-title-threadsafe.c [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    ASSERT(uv_set_process_title(titles[0]) == 0);
+    ASSERT(uv_set_process_title(titles[1]) == 0);
+    ASSERT(uv_set_process_title(titles[2]) == 0);
+    ASSERT(uv_set_process_title(titles[3]) == 0);
+  }
+}
+
+
+TEST_IMPL(process_title_threadsafe) {
+  uv_thread_t setter_threads[NUM_SETTER_THREADS];
```
Comment (bnoordhuis) [inline test/test-process-title-threadsafe.c]: I'd write this as `uv_thread_t setter_threads[4];` and use `ARRAY_SIZE(setter_threads)` below.

**Comment by bnoordhuis** [comment]
> I noticed the warning about the functions not being thread safe is only present in the v1.x branch docs and not master which I'm making the PR against. Is there anything to be done in this case? Should I be making this PR against v1.x?

Master is quite a bit behind v1.x right now, see #1630, that's why it's not in the docs there.  v1.x is the right branch to target for anything that is not semver-major.

Aside: looks like we're missing a couple NULL check in some of the BSD `uv_set_process_title()` implementations.  Something for a second PR?

**Comment by mtharrison** [comment]
Thanks for the feedback, I'm going to close this and reopen a PR on v1.x branch and updates the docs.

**Pull request closed without merging**